### ISSUE TITLE
os/user: doesn't work with musl libc

### ISSUE THREAD

**Issue by dominikh** [golang/go#11319]
When building Go on Linux with musl instead of glibc, os/user doesn't work:

$ go test os/user 
--- FAIL: TestCurrent (0.00s)
	user_test.go:23: Current: user: unreasonable _SC_GETPW_R_SIZE_MAX of -1
--- FAIL: TestLookup (0.00s)
	user_test.go:64: Current: user: unreasonable _SC_GETPW_R_SIZE_MAX of -1
--- FAIL: TestLookupId (0.00s)
	user_test.go:82: Current: user: unreasonable _SC_GETPW_R_SIZE_MAX of -1

That's because Go assumes that on Linux, _SC_GETPW_R_SIZE_MAX will always be >0, which isn't true.

I have a fix for this ready. Is this something for Go 1.5, or should I wait till 1.6?

**Comment by ianlancetaylor** [comment]
At this point, let's wait for 1.6.

**Comment by dominikh** [comment]
I figured as much.

**Comment by somasis** [comment]
I'm still hitting this issue with musl-1.1.12, and go-1.5.2, bootstrapping from go-1.4.3.

```
# os/user
os/user/lookup_unix.go:70: cannot use C.ulong(bufSize) (type C.ulong) as type C.size_t in argument to _Cfunc__CMalloc
os/user/lookup_unix.go:83: cannot use C.ulong(bufSize) (type C.ulong) as type C.size_t in argument to _Cfunc_mygetpwnam_r
os/user/lookup_unix.go:98: cannot use C.ulong(bufSize) (type C.ulong) as type C.size_t in argument to _Cfunc_mygetpwuid_r
```

**Comment by bradfitz** [comment]
It'll be in Go 1.6. 1.5.x is only for critical things, not nice to haves.

**Comment by dominikh** [comment]
What @somasis is reporting looks like a separate issue to me.

**Comment by somasis** [comment]
I should have said, I actually patched go-1.5.2 with the commit related to this issue, and still hit that error.

**Comment by Thomasdezeeuw** [comment]
@somasis  Have you tried tip (latest development version)?

**Comment by somasis** [comment]
@Thomasdezeeuw Yes, I get the same error.

**Comment by somasis** [comment]
I managed to get it to build with `CGO_ENABLED=0` passed to the final `./make.bash`. That doesn't sound like the right solution, though.

**Comment by somasis** [comment]
...which then fails to build programs such as syncthing, because of the exact same error.

**Comment by andlabs** [comment]
@somasis the compile errors you pasted are unrelated to this issue and are possibly a bug in cgo. If there isn't a similar one, you should file a new one.

**Comment by williamh** [comment]
Is this still an issue in Go 1.6? I'm hearing concerns about musl in Gentoo [1].

[1] https://bugs.gentoo.org/show_bug.cgi?id=575510#C1

**Comment by ianlancetaylor** [comment]
@williamh This issue is closed as fixed in 1.6.  Please take any questions about it to a forum; see https://golang.org/wiki/Questions .  Thanks.
